File: go/algorithm/advector1d/central_difference.go
package advector1d

import (
	"github.com/maxkuzn/advection-and-coagulation/internal/cell"
	"github.com/maxkuzn/advection-and-coagulation/internal/field1d"
)

type centralDifference struct {
	sigma cell.FloatType
}

func NewCentralDifference(sigma float64) *centralDifference {
	return &centralDifference{
		sigma: cell.FloatType(sigma),
	}
}

func (a *centralDifference) Process(field, buff field1d.Field) (field1d.Field, field1d.Field) {
	fieldIn := field
	fieldOut := buff

	l := field.Len()

	fieldOut.SetCell(
		0,
		cell.Sum3(
			1, fieldIn.Cell(0),
			-a.sigma/2, fieldIn.Cell(1),
			0, fieldIn.Cell(l-1),
		),
	)

	fieldOut.SetCell(
		l-1,
		cell.Sum3(
			1, fieldIn.Cell(l-1),
			0, fieldIn.Cell(0),
			a.sigma/2, fieldIn.Cell(l-2),
		),
	)

	for x := 1; x < l-1; x++ {
		fieldOut.SetCell(
			x,
			cell.Sum3(
				1, fieldIn.Cell(x),
				-a.sigma/2, fieldIn.Cell(x+1),
				a.sigma/2, fieldIn.Cell(x-1),
			),
		)
	}

	return fieldOut, fieldIn
}

File: cpp/library/algorithm/advector1d/central_difference.cpp
#include "central_difference.h"

namespace advection {

std::pair<Field1D*, Field1D*> CentralDifference::Process(Field1D* field, Field1D* buff) {
	Field1D& field_in = *field;
	Field1D& field_out = *buff;

	auto s = field_in.Size();

	field_out[0].AssignSum(
			1, field_in[0],
			-sigma_/2, field_in[1],
			0, field_in[s-1]
	);

	field_out[s-1].AssignSum(
			1, field_in[s-1],
			0, field_in[0],
			sigma_/2, field_in[s-2]
	);

	for (size_t x = 1; x < s-1; x++) {
		field_out[x].AssignSum(
				1, field_in[x],
				-sigma_/2, field_in[x+1],
				sigma_/2, field_in[x-1]
		);
	}


	return std::make_pair(&field_out, &field_in);
}

};  // namespace advection
